brave/src/main/java/brave/internal/codec/EntrySplitter.java

```diff
@@ -0,0 +1,315 @@
+
+import brave.internal.Platform;
+
+/**
+ * Splits a character sequence that's in a delimited string, optionally trimming optional whitespace
+ * (<a href="https://httpwg.org/specs/rfc7230.html#rfc.section.3.2">OWS</a>) before or after
+ * delimiters.
+ *
+ * <p>This is intended to be initialized as a constant, as doing so per-request will add
+ * unnecessary overhead.
+ */
+public final class EntrySplitter {
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static final class Builder {
+    int maxEntries = Integer.MAX_VALUE;
+    char entrySeparator = ',', keyValueSeparator = '=';
+    boolean trimOWSAroundEntrySeparator = true, trimOWSAroundKeyValueSeparator = true;
+    boolean keyValueSeparatorRequired = true, shouldThrow = false;
+
+    /**
+     * When set, {@link Handler} will be called maximum {@code maxEntries} times per parse. After
+     * that, {@link #parse(Handler, Object, String)} returns false or throws an exception, based on
+     * {@link #shouldThrow(boolean)}. Default: {@link Integer#MAX_VALUE}.
+     *
+     * <p>This is used to implement strict format constraints. For example, above 32 entries is
+     * malformed. This is separate from any capacity constraints of the {@link Handler}, which may
+     * be smaller than this amount.
+     */
+    public Builder maxEntries(int maxEntries) {
+      if (maxEntries <= 0) throw new IllegalArgumentException("maxEntries <= 0");
+      this.maxEntries = maxEntries;
+      return this;
+    }
+
+    /**
+     * The separator to use between entries. For example, the input "k1=v1,k2=v2", should have
+     * {@code entrySeparator} ','. Default: ','
+     *
+     * @see #keyValueSeparator(char)
+     */
+    public Builder entrySeparator(char entrySeparator) {
+      if (entrySeparator == 0) throw new IllegalArgumentException("entrySeparator == 0");
+      if (entrySeparator == keyValueSeparator) {
+        throw new IllegalArgumentException("entrySeparator == keyValueSeparator");
+      }
+      this.entrySeparator = entrySeparator;
+      return this;
+    }
+
+    /**
+     * The separator to use between a key and value. For example, the input "k1=v1,k2=v2" should
+     * have {@code keyValueSeparator} '='. Default: '='
+     *
+     * <p><em>Note:</em> Only the first {@code keyValueSeparator} identifies the end of the key
+     * until the next {@link #entrySeparator(char)}. This means values can include the {@code
+     * keyValueSeparator} character.
+     *
+     * <p>For example, the input "condition=animal=cat" with {@code keyValueSeparator} '=' parses
+     * {@code [("condition", "animal=cat")]}
+     *
+     * @see #keyValueSeparator(char)
+     */
+    public Builder keyValueSeparator(char keyValueSeparator) {
+      if (keyValueSeparator == 0) throw new IllegalArgumentException("keyValueSeparator == 0");
+      if (keyValueSeparator == entrySeparator) {
+        throw new IllegalArgumentException("keyValueSeparator == entrySeparator");
```

Comment:
Validating in the setter is nicer but I guess it means we should default the values to `0` and replace with the actual defaults in `build`. Otherwise, it requires hoops to do the horrible, but not inconceivable, using `=` as entry separator and `,` as keyvalue separator.

Reply:
I'm convinced deferring is better :D